src/coreclr/tests/src/JIT/HardwareIntrinsics/Arm/Shared/Helpers.tt

```diff
@@ -516,6 +516,39 @@ namespace JIT.HardwareIntrinsics.Arm
             return saturated;
         }
 
+        public static <#= type.name #> ExtractNarrowingSaturate(<#= type.wide #> op1)
+        {
+            <#= type.name #> result;
+
+            SignedSatQ(op1, out result);
+
+            return result;
+        }
+
+        public static <#= type.name #> ExtractNarrowingSaturateUpper(<#= type.name #>[] op1, <#= type.wide #>[] op2, int i) => i < op1.Length ? op1[i] : ExtractNarrowingSaturate(op2[i - op1.Length]);
+
+        public static <#= type.unsigned #> ExtractNarrowingSaturate(<#= type.wideUnsigned #> op1)
+        {
+            <#= type.unsigned #> result;
+
+            UnsignedSatQ(op1, out result);
+
+            return result;
+        }
+
+        public static <#= type.unsigned #> ExtractNarrowingSaturateUpper(<#= type.unsigned #>[] op1, <#= type.wideUnsigned #>[] op2, int i) => i < op1.Length ? op1[i] : ExtractNarrowingSaturate(op2[i - op1.Length]);
+
+        public static <#= type.unsigned #> ExtractNarrowingSaturateUnsigned(<#= type.wide #> op1)
+        {
+            <#= type.unsigned #> result;
+
+            SignedSatQ(op1, out result);
```

Comment:
I'm confused by this helper method, and the one above; they seem backward to me. As I understand it, `uqxtn` (which is used for `ExtractNarrowingSaturateLower` with unsigned baseTypes) reads an unsigned value and does signed saturation, while this variant which translates to `sqxtun` should be reading the value as signed and doing unsigned saturation.Can you help me understand this?

Reply:
I have the same question, this should be `ExtractNarrowingSaturateUnsigned` which is `SQXTUN`, the operation for which is:
![image](https://user-images.githubusercontent.com/10487869/85873397-e93f6200-b785-11ea-85ca-6583f4a11b0f.png)

Reply:
I see how I caused this confusion by wrongly naming UnsignedSatQ(SignedWideType val, out UnsignedNarrowType result) as SignedSatQ. The underlying implementation was correct, so the tests passed - the helper name was wrong :-)

In fact, I found that we only need the following three helpers:
- SignedSatQ(SignedWideType val, out SignedNarrowType result)
- UnsignedSatQ(SignedWideType val, out UnsignedNarrowType result)
- UnsignedSatQ(UnsignedWideType val, out UnsignedNarrowType result)

I added SatQ helper as defined in the Arm specification and used it instead of SignedSatQ/UnsignedSatQ to make the implementation looking closer to the *"gold standard"*.

However, since we decided to provide right logical shifts overloads for signed types - I had to extend SatQ helper operating on signed arguments in the way allowing reinterpretation of the incoming signed values as unsigned that helps avoiding doing multiple casting in different helpers.

I pushed the change to Helpers.tt Helpers.cs - it makes sense to look at the former only.